tests/qat/test_gates.py

```diff
@@ -0,0 +1,238 @@
+from itertools import product
+
+import numpy as np
+import pytest
+
+from qat.purr.integrations.qasm import (
+    Qasm2Parser,
+    Qasm3Parser,
+    Qasm3ParserBase,
+    QasmContext,
+)
+
+from tests.qat.utils.matrix_builder import Gates, get_default_matrix_hardware
+
+
+def assert_same_up_to_phase(gate1, gate2):
+    gate = np.conj(np.transpose(gate1)) @ gate2
+    gate /= gate[0, 0]
+    assert np.isclose(gate, np.eye(np.shape(gate1)[0])).all()
+
+
+def single_gate_list():
+    func_to_gates = {
+        "SX": ["SX", Gates.sx(), ()],
+        "SXdg": ["SXdg", np.conj(np.transpose(Gates.sx())), ()],
+        "T": ["T", Gates.t(), ()],
+        "Tdg": ["Tdg", np.conj(np.transpose(Gates.t())), ()],
+        "hadamard": ["had", Gates.h(), ()],
+    }
+
+    thetas = [0.0, np.pi / 2, np.pi, -np.pi / 2, 0.321, -1.58]
+    for theta in thetas:
+        func_to_gates[f"Rx({theta})"] = ["X", Gates.rx(theta), (theta,)]
+        func_to_gates[f"Ry({theta})"] = ["Y", Gates.ry(theta), (theta,)]
+        func_to_gates[f"Rz({theta})"] = ["Z", Gates.rz(theta), (theta,)]
+
+    params = product(thetas, thetas, thetas)
+    for args in params:
+        func_to_gates[f"U({args[0]}, {args[1]}, {args[2]})"] = ["U", Gates.U(*args), args]
+
+    return func_to_gates
+
+
+class TestSingleGates:
+
+    model = get_default_matrix_hardware(1)
```

Comment:
Just to verify it's still working on multi-qubit systems, can we also have tests for `n_qubits > 1`?

Reply:
So I somewhat confusingly placed a test that does this for `n_qubits = 2` under `TestDoubleGates`, but this could definitely be more robust (e.g. I'm only doing this for single qubit gates, we could also verify `n_qubits > 2` for two-qubit gates) :sweat_smile: 

I'll add some more, thanks! :smile: